src/mpl/src/gavl/mpl_gavl.c

```diff
@@ -227,10 +435,16 @@ int MPL_gavl_tree_insert(MPL_gavl_tree_t gavl_tree, const void *addr, uintptr_t
                     direction = SEARCH_RIGHT;
                 }
             } else {
-                /* we cannot insert the duplicate buffer */
-                gavl_tree_iptr->gavl_free_fn((void *) node_ptr->val);
+                mpl_err =
+                    gavl_tree_delete_node(gavl_tree, (void *) cur_node->addr, cur_node->len,
+                                          &cur_node);
+                MPL_ERR_CHECK(mpl_err);
+
+                mpl_err = gavl_tree_node_merge(node_ptr, cur_node);
+                MPL_ERR_CHECK(mpl_err);
                 MPL_free(node_ptr);
-                goto fn_exit;
+                node_ptr = cur_node;
+                goto restart;
             }
```

Comment:
Why do we go restart? Does it mean we search from the tree root again?

Reply:
yes, please see my previous comment